app/models/spree/tax_rate.rb

```diff
@@ -30,16 +32,32 @@ def self.match(order)
       return [] if order.distributor && !order.distributor.charges_sales_tax
       return [] unless order.tax_zone
 
-      all.select do |rate|
-        rate.zone == order.tax_zone || rate.zone.contains?(order.tax_zone) || rate.zone.default_tax
+      all.includes(zone: { zone_members: :zoneable }).load.select do |rate|
+        rate.potentially_applicable?(order.tax_zone)
       end
     end
 
-    def self.adjust(order)
-      order.all_adjustments.tax.destroy_all
+    def self.adjust(order, items)
+      applicable_rates = match(order)
+      applicable_tax_categories = applicable_rates.map(&:tax_category)
+
+      relevant_items, non_relevant_items = items.partition do |item|
+        applicable_tax_categories.include?(item.tax_category)
+      end
 
-      match(order).each do |rate|
-        rate.adjust(order)
+      relevant_items.each do |item|
+        item.adjustments.tax.delete_all
+        relevant_rates = applicable_rates.select { |rate| rate.tax_category == item.tax_category }
+        relevant_rates.each do |rate|
+          rate.adjust(order, item)
+        end
+      end
+
+      non_relevant_items.each do |item|
+        if item.adjustments.tax.present?
+          item.adjustments.tax.delete_all
+          Spree::ItemAdjustments.new(item).update
+        end
```

Comment:
this class is so complicated. I think this method (adjust) could go directly to a service, by itself with some good unit testing.